EventSourcing.Core/Records/RecordConverter.cs

```diff
@@ -0,0 +1,144 @@
+    _migrators = (options?.MigratorTypes ?? AssemblyMigratorTypes)
+      .Select(type => Activator.CreateInstance(type) as IRecordMigrator)
+      .ToDictionary(migrator => migrator!.Source.Name, migrator => migrator);
+
+    ValidateMigrators();
+  }
+
+  /// <summary>
+  /// Use <see cref="RecordConverter{TRecord}"/> for all Types inheriting from <see cref="Record"/>
+  /// </summary>
+  /// <param name="typeToConvert">Type to Convert</param>
+  public override bool CanConvert(Type typeToConvert) => typeof(Record).IsAssignableFrom(typeToConvert);
+
+  /// <summary>
+  /// Serialize Record
+  /// </summary>
+  public override void Write(Utf8JsonWriter writer, TRecord value, JsonSerializerOptions options)
+  {
+    var type = GetRecordType(value.Type);
+    var record = Validate(value, type);
+    JsonSerializer.Serialize(writer, record, type);
+  }
+
+  /// <summary>
+  /// Deserialize Record
+  /// </summary>
+  /// <exception cref="JsonException">Thrown when <see cref="Record"/> type cannot be found.</exception>
+  public override TRecord Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+  {
+    var type = GetRecordType(reader);
+    var record = JsonSerializer.Deserialize(ref reader, type) as TRecord;
+    return Migrate(Validate(record, type));
+  }
+
+  private Type GetRecordType(Utf8JsonReader reader)
+  {
+    // Get Record.Type String from Json
+    var typeString = JsonSerializer.Deserialize<RecordType>(ref reader)?.Type ??
+                     
+       // Throw Exception when json has no "Type" Property
+       throw new RecordValidationException($"Error while extracting record type string. Does the JSON contain a {nameof(Record.Type)} field?");
+
+    return GetRecordType(typeString);
+  }
+
+  private Type GetRecordType(string typeString)
+  {
+    // Get actual Record Type from Dictionary
+    if (!_recordTypes.TryGetValue(typeString, out var type))
+      
+      // Throw Exception when Record Type is not found in Assembly or RecordConverterOptions
+      throw new InvalidOperationException($"Record with Type '{typeString}' not found");
+
+    return type;
+  }
+
+  private TRecord Validate(TRecord record, Type type)
+  {
+    var missing = _nonNullableRecordProperties[type]
+      .Where(property => property.GetValue(record) == null)
+      .Select(property => property.Name)
+      .ToList();
+    
+    if (missing.Count > 0)
+      throw new RecordValidationException($"Error validating {type} with RecordId '{record.RecordId}'.\n" +
+        $"One ore more non-nullable properties missing or null: {string.Join(", ", missing.Select(property => $"{type.Name}.{property}"))}");
+
+    return record;
+  }
+
+  private TRecord Migrate(TRecord record)
+  {
+    while (_migrators.TryGetValue(record.Type, out var migrator))
+      record = (TRecord) migrator.Convert(record);
+
+    return record;
+  }
+  
+  private void ValidateMigrators()
+  {
+    foreach (var m in _migrators.Values)
```

Comment:
To nitpick my own foreach loop: Each chain is now being checked as many times as there are elements in the chain. We can change it to a while loop en using the visited list to make sure each element is only checked once.